### PR TITLE
Handle non scalar values

### PR DESCRIPTION
**Pull request by kmadisa**
When writing to a _SPECTRUM_ type attribute, the data structure passed on to the write method is converted into an _numpy ndarray_ which was not handled. This was causing an error when it was trying to type cast an `numpy.ndarray` into an `int`.

The other issue is that the simulated quantity values were not being correctly reset to the default values. 
`simulated_quantity.attr = adjustable_val` was creating a new quantity attribute, `attr`, instead of assigning the value to the existing quantity attributes e.g. `(last_val, mean, std_dev, max_slew_rate, min_bound, max_bound)`.

This fixes the unit tests in [katproxy/pull/539](https://github.com/ska-sa/katproxy/pull/539)

*Definition of Done Checklist*

- [x] Code meets our [python style guidelines](https://docs.google.com/document/d/1aZoIyR9tz5rCWr2qJKuMTmKp2IzHlFjrCFrpDDHFypM/edit?usp=sharing)?
- [x] Unit tested (coded, passed, included)?
- [x] Requested at least 2 reviewers?
- [x] Commented code, particularly in hard-to-understand areas?
- [x] Made corresponding changes to the documentation (e.g. Python documentation, System Engineering Documentation, version description updates, README file, etc)?

JIRA: [MT-38](https://skaafrica.atlassian.net/browse/MT-38)

### CODE REVIEW

File: tango_simlib/tango_sim_generator.py [hunk 1]
```diff
@@ -297,8 +298,18 @@ def _reset_to_default_state(self):
...
+                        elif attr == 'mean':
+                            simulated_quantity.mean = adjustable_val
+                        elif attr == 'std_dev':
+                            simulated_quantity.std_dev = adjustable_val
+                        elif attr == 'max_slew_rate':
+                            simulated_quantity.max_slew_rate = adjustable_val
+                        elif attr == 'min_bound':
+                            simulated_quantity.min_bound = adjustable_val
+                        elif attr == 'max_bound':
+                            simulated_quantity.max_bound = adjustable_val
```
Comment (ajoubertza) [inline tango_simlib/tango_sim_generator.py]: Should we raise an exception if it is none of these options?  Also, code could be a little shorter if we do a `hasattr` check and then a  `setattr` to work more generically.

File: tango_simlib/tests/test_sim_xmi_parser.py [hunk 2]
```diff
@@ -658,3 +658,15 @@ def test_spectrum_attribute_properties(self):
...
+        """Test that the Spectrum writable attribute can be set correctly."""
+        timestamp = 0.0
+        az = 0.0
+        el = 0.0
+        az_speed = 0.0
+        el_speed = 0.0
+        az_accl = 0.0
+        el_accl = 0.0
+        self.assertEqual(self.device.desiredPosition, [timestamp, az, el, az_speed,
+                                                       el_speed, az_accl, el_accl])
```
Comment (ajoubertza) [inline tango_simlib/tests/test_sim_xmi_parser.py]: Are these tests still a work in progress?  Attribute is not being set.

Comment (kmadisa) [reply tango_simlib/tests/test_sim_xmi_parser.py]: It is a work in progress.

File: tango_simlib/tango_sim_generator.py (hunk 1 above)
Comment (kmadisa) [reply tango_simlib/tango_sim_generator.py]: We do not need to raise an exception. We get them from https://github.com/ska-sa/tango-simlib/blob/fa6532247d2f4f8e633c7f863fd5cef05eb9d623/tango_simlib/tango_sim_generator.py#L248-L249

and loop through the list.

File: tango_simlib/tango_sim_generator.py [hunk 3]
```diff
@@ -248,7 +263,7 @@ def _reset_to_default_state(self):
                 adjustable_attrs = simulated_quantity.adjustable_attributes
                 for attr in adjustable_attrs:
                     if attr == 'last_update_time':
-                        simulated_quantity.attr = model_time
+                        setattr(simulated_quantity, attr, model_time)
```
Comment (ajoubertza) [inline tango_simlib/tango_sim_generator.py]: In this case there is only one attribute, so the generic `setattr` mechanism isn't that useful.  At least the error of writing to the wrong attribute has been fixed!

File: tango_simlib/tango_sim_generator.py [hunk 4]
```diff
@@ -292,13 +307,13 @@ def _reset_to_default_state(self):
                                     adjustable_val = val_type(adjustable_val)
                             else:
                                 if attr == 'last_val':
-                                    simulated_quantity.attr = float(sim_quantity_meta_info['mean'])
+                                    setattr(simulated_quantity, attr,
+                                            float(sim_quantity_meta_info['mean']))
```
Comment (ajoubertza) [inline tango_simlib/tango_sim_generator.py]: Similar comment to above, could be `simulated_quantity.last_val = ...`

File: tango_simlib/tango_sim_generator.py [hunk 5]
```diff
@@ -350,7 +365,14 @@ def initialize_dynamic_attributes(self):
...
                                 "No setter function for " + prop + " property")
                     attr.set_default_properties(attr_props)
-                    self.add_attribute(attr, self.read_attributes)
+
+                    if rw_type == AttrWriteType.READ:
+                        self.add_attribute(attr, self.read_attributes)
+                    elif rw_type in (AttrWriteType.READ_WRITE,
+                                     AttrWriteType.READ_WITH_WRITE, AttrWriteType.WRITE):
+                        self.add_attribute(
+                            attr, self.read_attributes, self.write_attributes)
```
Comment (ajoubertza) [inline tango_simlib/tango_sim_generator.py]: If it is `AttrWriteType.WRITE`, then should it get `read_attributes`?

File: tango_simlib/tango_sim_generator.py [hunk 6]
```diff
@@ -62,6 +63,20 @@ def read_attributes(self, attr):
...
+        Parameters
+        ----------
+        attr : PyTango.DevAttr
+            The attribute to write to.
+
+        """
+        name = attr.get_name()
+        data = attr.get_write_value()
+        self.info_stream("Writing attribute {} with value: {}".format(name, data))
+        self.model.sim_quantities[name].set_val(data, self.model.time_func())
```
[the remark this answers is not in the document]
Comment (kmadisa) [reply tango_simlib/tango_sim_generator.py]: That is correct. The `model.update` method updates the `model.quantity_state` with the latest values set in the `sim_quantities`.
https://github.com/ska-sa/tango-simlib/blob/fa6532247d2f4f8e633c7f863fd5cef05eb9d623/tango_simlib/model.py#L107-L114

File: tango_simlib/tango_sim_generator.py (hunk 5 above)
Comment (kmadisa) [reply tango_simlib/tango_sim_generator.py]: No it shouldn't. I have removed it.

**Pull request merged**